js/renovation/utils/plugin/context.ts

```diff
@@ -36,10 +36,46 @@ export class PluginGetter<T> extends PluginEntity<T, GetterStoreValue<T>> {
   }
 }
 
+export type SelectorFunc<R> = (...args) => R;
+
+export class PluginSelector<R> extends PluginEntity<R, R> {
+  readonly deps: PluginEntity<unknown, unknown>[];
+
+  readonly func: SelectorFunc<R>;
+
+  constructor(deps: PluginEntity<unknown, unknown>[], func: SelectorFunc<R>) {
+    super();
+    this.deps = deps;
+    this.func = func;
+  }
+}
+
 export function createValue<T>(): PluginEntity<T, T> {
   return new PluginEntity<T, T>();
 }
 
+// export function createSelector<A, R>(
+//   deps: [PluginEntity<A, A>],
+//   func: (a: A) => R
+// ): PluginSelector<R>;
+
+// export function createSelector<A, B, R>(
+//   deps: [PluginEntity<A, A>, PluginEntity<B, B>],
+//   func: (a: A, b: B) => R
+// ): PluginSelector<R>;
+
+// export function createSelector<A, B, C, R>(
+//   deps: [PluginEntity<A, A>, PluginEntity<B, B>, PluginEntity<C, C>],
+//   func: (a: A, b: B, c: C) => R
+// ): PluginSelector<R>;
+
+export function createSelector<R>(
+  deps: PluginEntity<unknown, unknown>[],
+  func: SelectorFunc<R>,
+): PluginSelector<R> {
+  return new PluginSelector<R>(deps, func);
+}
+
```

Comment:
There is ts playground with variadic generics, but it doesn't infer types of func's params

[link](https://www.typescriptlang.org/play?#code/LAKAlgdgLgpgTgMwIYGMYAIAKAbArgc0gFFowoBPAHgEEA+dAb1HRfQGcB7AWxgDUk8MAFzpqoAL6hIsRKgw4CkAMoxsMFFA5wa9BpJChQAE3XYkcDAlwQNYDhHQoLSWCrUatlACroYAD1gIIzZ0JAhyAG0AXQAadAAlWgAKZlYTAAc2EQZ0CIBpdEh0AGsYcg4EdC8okQVCCBIoMiovfKj6cRjUlisbESSAOiHzfCyqgEp0AF56eNBx2rx6t3VNbUSAbkMQNSgsJHwYAGEOayhFxQbSCkoIXC4AI3h6KcZOHn5BEQBGcS2dmB7JRQOCQfCfXDCLBLYjXKhsEFgl5vbh8ASQkQAciQbCMmL+2wA9IT0N8BugAOpgbDYRzcdLUjAAdzIAAt0KD8Ky9hR0jA2KAnDAXDAVh44CkQKxcsDQRBweiYHFMAdjqdoLFuugkgi5QqvuxEfK4ulVScziI7o94JMZoxttL9OMiSSAEzkin2TF7FD0xnoFlQdlMuD2fDoXn8wXOVyqVZaSXSiKysEQpX7Q7mjVdKWsHVG-UY9BWp5wE1m9XnYv3Uu23QO1hOl3oADMHuptOQNIeqGKEY46GsdgcYEqkZC5gwbGKYHSfKM0eFsfca0TrGTBbTyorZ01uZY+b1W-Qpszlbr9v30vQWqbBhAQA)

Reply:
another one, with infering: [link](https://www.typescriptlang.org/play?#code/LAKAlgdgLgpgTgMwIYGMYAIAKAbArgc0gFFowoBPAHgEEA+dAb1HRfQGcB7AWxgDUk8MAFzpqoAL6hIsRKgw4CkAMoxsMFFA5wa9BpJChQAE3XYkcDAlwQNYDhHQoLSWCrUatlACroYAD1gIIzZ0JAhyAG0AXQAadAAlWgAKZlYrGxEkgDoc83w2ES8ASnQAXnp4mNSWHKyTAAcCxnQIgGl0SHQAaxhyDgR0LyiRBUIIEigyKi82qPpxKpAikbwxt3VNbUSAbkMQNSgsJHwYAGEOaygVxXHSCkoIXC4AI3h6UsZOHn5BEQBGcS7fYwQ5KKBwSD4H64YRYVbEO5UNjgyHvT7cPgCGEiADkSDYRhxgL2AHoSeg-ll0AB1MDYbCObj1OkYADuZAAFugIfgOYcKPUYGxQE4YC4YOsPHAUiBWOgksiIRAoVjYYrIXF6sczhdoCJHi94CVyow9nKFtV0GClSrBHFMNrzpdFkVSeSAExU6n2HGHFBMlnodlQLmsuD2fDoAVCkXOVyqDZaGVyhUo5XQ2EG15wTWO3VXdBZo1lXRm1gW2Wsa2QjP2vPO0CugwgMnoADMVOjHQgCHgkNjYvj7k2ydYqZttfQWpOTugxtLlblLEtFbl1fTqrrM-zLr2QA)

Reply:
The both suggested solutions are not perfect. So, I keep current approach.